telepresence/startup.py

```diff
@@ -31,23 +31,34 @@ def kubectl_or_oc(server: str) -> str:
 
     :param server: The URL of the cluster API server.
     """
-    if which("oc") is None:
-        return "kubectl"
+    kubectl = "kubectl"
+    oc = "oc"
+
+    if which(oc) is None:
+        return kubectl
     # We've got oc, and possibly kubectl as well. We only want oc for OpenShift
-    # servers, so check for an OpenShift API endpoint:
+    # servers, so check for an OpenShift API endpoint exposing users
+    # (it's also used by oc whoami command):
     ctx = ssl.create_default_context()
     ctx.check_hostname = False
     ctx.verify_mode = ssl.CERT_NONE
     try:
-        with urlopen(server + "/apis", context=ctx) as response:
+        with urlopen(
+            server + "/apis/user.openshift.io/v1/users/", context=ctx
+        ) as response:
             api_group_list = str(response.read())
-    except (URLError, HTTPError):
-        return "kubectl"
+    except HTTPError as err:
+        if err.code == 403:
+            return oc
```

Comment:
Why is this special-cased? I'm tripping on this now with my old-ish GKE cluster.

```console
$ oc whoami
error: No Auth Provider found for name "gcp"
```

Reply:
Sorry @ark3, you caught me during my holidays with this one. Can you provide me some more details about the GKE cluster which is causing the problem?

Reply:
The general idea is/was - if we have both `oc` and `kubectl` on the `$PATH` then the assumption is we most likely want to use OpenShift tooling and we connect to OpenShift cluster. If we call for api groups and getting `unauthorized` for the Openshift-specific endpoint, then we stick with OpenShift anyway.

Maybe we could also provide an optional flag for the cli (e.g. `--cluster-type`) where user can specifically say what is the actual vendor of the target cluster. If you think that would make sense I can open an issue and I'm happy to work on it.

Reply:
GKE appears to return `unauthorized` for _all_ endpoints if it needs to grab a new access token or whatever using the auth provider helper. The code ends up thinking the cluster is OpenShift when really it's just a general failure to auth to the cluster. The `oc` error above is actually not the point, though it implies that `oc` doesn't know how to use Google Cloud's auth helper. But yeah, sorry to misdirect you with that.

Anyhow, I replaced your code (which is still there, but not called) with a different approach. Now Tel calls `kubectl api-versions` (or `oc api-versions` if `kubectl` is not available) and searches the output for "openshift.io". The implementation details are somewhat hackish (bad error handling in some cases) but the general approach seems sound. What do you think?

I ended up releasing this in 0.103. Can you please check it out? I hope I didn't break what you were fixing with your PR.